Moved "Installing Dependencies" guide to main docset (#374)
astronomer/docs: enterprise/v0.25/10_certified-image/02_install-packages.md

---
title: "Install Packages to Astronomer Certified"
navTitle: "Install Packages"
description: "Learn how to install OS-level and Python-level packages on Airflow."
---

## Overview

> By default, the Astronomer Certified Docker image is distributed with a collection of pre-installed Python and OS-level packages to help users integrate with popular applications. Python-level packages are dependencies that Airflow uses, while OS-level packages are dependencies required for the underlying Debian OS. For the full list of built-in packages, read [Astronomer Certified Image Architecture].

Reviewer: The link is missing here, right?

Author: Yep, it will be added after we've merged all docs to the pre-publication branch.

Depending on your use case and distribution of Astronomer Certified, you might want to install additional packages to your environment. This guide provides steps for installing dependencies to both Astronomer's Docker image and Python wheel.

## Install Packages to the Docker Image

If you use the Astronomer Certified Docker image to run Airflow, you can install packages directly onto your image via your `Dockerfile`. To install OS-level packages, you can specify them using a `RUN` directive with `apt-get`. For example, the following `Dockerfile` would install `your-os-package` on the image:

```
FROM quay.io/astronomer/ap-airflow:2.1.0-buster-onbuild
RUN apt-get update \
  && apt-get install -y --no-install-recommends \
         <your-os-package> \
  && apt-get autoremove -yqq --purge \
  && apt-get clean \
  && rm -rf /var/lib/apt/lists/*
```

To install a specific version of an OS-level package, use the following format for your installation command:

```
apt-get install <your-os-package>=<version> -V
```

To install a Python-level package, specify the package using a `RUN` directive with `pip install` instead. For example:

```

> FROM: quay.io/astronomer/ap-airflow:latest-onbuild

Reviewer 2: ```suggestion
FROM quay.io/astronomer/ap-airflow:2.1.0-buster-onbuild
```

> RUN pip install --no-cache-dir --user <your-python-package>

Reviewer: python level packages can also be install using `requirements.txt` as explained in https://www.astronomer.io/docs/cloud/stable/develop/customize-image#add-python-and-os-level-packages

Reviewer: oh I see you have this in the section below where you talk about astro CLI, should we swap the order?

Author: I think our plan is to focus on this doc as if someone isn't using the Astronomer platform or CLI to run the image. I'm about to remove the section below and instead link out to Customize Image as the primary source of truth for using the Astro CLI for image customization.

---
title: "Install Packages to Astronomer Certified"
navTitle: "Install Packages"
description: "Learn how to install OS-level and Python-level packages on Airflow."
---

## Overview

> By default, the Astronomer Certified Docker image is distributed with Python and OS-level packages to help users integrate with popular applications. Python-level packages are dependencies which Airflow uses, while OS-level packages are dependencies which the image's underlying Debian OS uses. For the full list of built-in packages, read [Astronomer Certified Image Architecture].

Reviewer 2: ```suggestion
By default, the Astronomer Certified Docker image is distributed with a collection of pre-installed Python and OS-level packages to help users integrate with popular applications. Python-level packages are dependencies that Airflow uses, while OS-level packages are dependencies required for the underlying Debian OS. For the full list of built-in packages, read [Astronomer Certified Image Architecture].
```

I think I liked the original wording here - "distributed with Python and OS-level packages" sounds confusing and makes me want to ask "as in some are pre-installed?"

> Depending on your use case and distribution of Astronomer Certified, you might want to install additional packages to your environment. This guide provides steps for installing additional dependencies to both Astronomer's Docker image and Python wheel.

Reviewer 2: ```suggestion
Depending on your use case and distribution of Astronomer Certified, you might want to install additional packages to your environment. This guide provides steps for installing dependencies to both Astronomer's Docker image and Python wheel.
```

## Install Packages to the Docker Image

> If you use the Astronomer Certified Docker image to run Airflow, you can install the packages directly onto your image via your Dockerfile. To install OS-level packages, you can specify them using a `RUN` directive with `apt-get`. For example, the following Dockerfile would install `your-package` on the image:

Reviewer 2: ```suggestion
If you use the Astronomer Certified Docker image to run Airflow, you can install packages directly onto your image via your `Dockerfile`. To install OS-level packages, you can specify them using a `RUN` directive with `apt-get`. For example, the following `Dockerfile` would install `your-os-package` on the image:
```

```

> FROM: quay.io/astronomer/ap-airflow:latest-onbuild

Reviewer 2: ```suggestion
FROM quay.io/astronomer/ap-airflow:2.1.0-buster-onbuild
```

Both are fine, but this is more representative I think. We don't really want to recommend that AC users depend on `latest-onbuild`

RUN apt-get update \

>   && apt-get install -y --no-install-recommends \
>          your-package \

Reviewer 2: ```suggestion
  && apt-get install -y --no-install-recommends \
         <your-os-package> \
```

  && apt-get autoremove -yqq --purge \
  && apt-get clean \

>   && rm -rf /var/lib/apt/lists/*

Reviewer 2: Do we know how to pin these packages to a particular version in the `RUN apt-get...` format? Might be helpful to add in here since we have instructions for that via Astro CLI

Author: This got a little more complicated than I expected! Could probably use @kaxil 's confirmation on what I wrote here. Also included a note below in case people are trying to deploy this image via the CLI, since that process for installing dependencies is different.

Reviewer: Yeah not-easily :) 

Btw they can also install this packages by adding name of package in `packages.txt`

```

To install a Python-level package, specify the package using a `RUN` directive with `pip install` instead. For example:

```
FROM: quay.io/astronomer/ap-airflow:latest-onbuild

> RUN pip install --no-cache-dir --user your-package

Reviewer 2: ```suggestion
RUN pip install --no-cache-dir --user <your-python-package>
```

    <package-name>==<version>
    ```

> ## Install Packages to the Python Wheel

Reviewer 2: Are we installing Python Packages TO the Python Wheel, or to the virtual machine? Is it moreso "Install Packages to a Virtual Machine with the AC Python Wheel" (or something like that)?

To build Python and OS-level packages into a machine running the Python wheel distribution of Astronomer Certified, run the following command:

```sh

> sudo -u astro ~astro/airflow-venv/bin/pip install --extra-index-url=https://pip.astronomer.io/simple/ 'astronomer-certified[<your-dependency>]==<airflow-version>.*'

Reviewer 2: ```suggestion
sudo -u astro ~astro/airflow-venv/bin/pip install --extra-index-url=https://pip.astronomer.io/simple/ 'astronomer-certified[<your-package>]==<airflow-version>.*'
```

```

You can also create your own Python packages and install them into your Airflow environment via a Python wheel, or you can configure an environment variable to automatically add the packages to your Airflow project directory. For more information on this setup, read the [Apache Airflow documentation on managing modules](http://apache-airflow-docs.s3-website.eu-central-1.amazonaws.com/docs/apache-airflow/latest/modules_management.html).

## Install Packages via the Astronomer CLI

If you set up a local Airflow environment as described in the [CLI Quickstart](docs/enterprise/v0.25/develop/cli-quickstart), you can add Python-level packages to your `requirements.txt` file and OS-level packages to your `packages.txt` file.

> To add the package, list the package name on its own line in the corresponding file. To install a particular version of any packages, pin that version to the package name in the format of `<package-name>==<package-version>`. To install Pymongo 3.7.2, for example, add the following to your `requirements.txt` file:

Reviewer 2: ```suggestion
To add the package, list the package name on its own line in the corresponding file. To install a particular version of any package, pin that version to the package name in the format of `<package-name>==<package-version>`. To install Pymongo 3.7.2, for example, add the following to your `requirements.txt` file:
```

Author: Now that we have this in the main docset, should we just link out to "Customize Images" here?

Reviewer 2: @jwitz Good call - yes, let's do that. Will defer to you on how much info you want to keep in here, but great call to reduce too much duplicate content
